netconf_client/ncclient.py

```diff
@@ -21,40 +24,258 @@
 )
 
 
+# Defines the scope for netconf traces
+logger = logging.getLogger("netconf_client_pretty")
+
+
+def pretty_xml(xml):
+    """Reformats a given string containing an XML document (for human readable output)"""
+
+    parser = etree.XMLParser(remove_blank_text=True)
+    tree = etree.fromstring(xml, parser)
+    return etree.tostring(tree, pretty_print=True).decode()
+
+
 class Manager:
-    """A helper class for performing common NETCONF operations
+    """A helper class for performing common NETCONF operations with pretty logging.
 
     This class attempts to be API compatible with the manager object
     from the original ncclient.
 
     This class is also a context manager and can be used with `with`
     statements to automatically close the underlying session.
 
+    NetConf requests and responses are logged using the 'netconf_client' scope.
+    The API functions, which are to be logged, can be configured (default: all but 'close_session').
+    Also the log level can be changed via API (logger.DEBUG is the default).
+
+    Each log entry shows a log ID (the peer's IP address as default).
+    Also a timestamp is printed for each request/response message and the round-trip delay
+    between request and its response is computed and displayed.
+
     :ivar float timeout: Duration in seconds to wait for a reply
 
     :ivar session: The underlying
                    :class:`netconf_client.session.Session` connected
                    to the server
+    :ivar str log_id: application-specific log ID (None as default)
+    :ivar int log_level: logging level (logging.DEBUG as default)
 
     """
 
-    def __init__(self, session, timeout=120):
+    def __init__(self, session, timeout=120, log_id=None):
         """Construct a new Manager object
 
         :param session: The low-level NETCONF session to use for requests
         :type session: :class:`netconf_client.session.Session`
 
         :param float timeout: Duration in seconds to wait for replies
+        :param string log_id: log ID string printed with each log entry,
+               the peer's IP address as default
         """
         self.timeout = timeout
         self.session = session
+        self.log_id = log_id
+        self.start_time = self._get_timestamp()
+        self.log_level = logging.DEBUG
+
+        self.logged_functions = {
+            "get": True,
+            "edit_config": True,
+            "get_config": True,
+            "copy_config": True,
+            "discard_changes": True,
+            "commit": True,
+            "lock": True,
+            "unlock": True,
+            "kill_session": True,
+            "create_subscription": True,
+            "validate": True,
+            "delete_config": True,
+            "dispatch": True,
+            "close_session": False,
+        }
 
     def __enter__(self):
         return self
 
     def __exit__(self, a, b, c):
         self.session.__exit__(a, b, c)
 
+    @staticmethod
+    def get_logger():
```

Comment:
This seems redundant. Either this method should be removed since the `logger` object is public already, or the `logger` object defined above should be private.

If you decide to keep this method, I'd suggest renaming it to just `logger()`.

Reply:
Ok.